docs/design/fabric-v2+/fpc-registration.puml

```diff
@@ -7,266 +7,200 @@
-      and, if epid, gets IAS report. 
-    - EnclavePersist: returns sealed state (important for security that this is called only after
-      EnclaveProof [which we might want to enforce using some state-machine?] )
-  end note
-  return ATTESTATION
+  return Credentials, Sealed_Credentials
 
-  |||
+  E_Manager -> E_Manager : store Sealed_Credentials
 
-  Manager1 -> Manager1 : mark ECC_ID as installed
-  note right
-    Manager1 maintains a list of
-    locally "installed/created" enclaves
-  end note
-  return ATTESTATION, ECC_S_PK / error
+  return result /' Admin -> E_Manager '/
 
   deactivate Admin
-  |||
+
+'  note over Admin1 #FFaa00
+'    **MST**: From an attestation generalization perspective i think there should be just a single
+'    call ""GetAttestationProof"" or alike rather than a separate place where you get quote
+'    and another where you get the registration report which requires some re-organization of calls.
+'    I think from a high-level there should be following three high-level calls either in
+'    ECC or ECCMAnager depending on above:
+'    - EnclaveCreate (key-gen & imprint of context, i.e., current create/binding flows
+'    - EnclaveProof: gets org endorsement-sig, creates report, then quote
+'      and, if epid, gets IAS report. 
+'    - EnclavePersist: returns sealed state (important for security that this is called only after
+'      EnclaveProof [which we might want to enforce using some state-machine?] )
+'  end note
+'  return ATTESTATION
 
 end
 
-group register attestation
+group Org-Enclave binding/certification (to be implemented)
+  Admin -> E_Manager ++: certify(CC_Id, Cert_Parameters)
+  E_Manager -> ECC1 ++: getCSR(Cert_Parameters)
+  ECC1 -> Enclave1 ++: getCSR(Cert_Parameters)
+  return csr /' ECC1 -> Enclave1 '/
+  return csr /' E_Manager -> ECC1 '/
+  E_Manager -> E_Manager : generate enclave certificate (certified by Org_1)
+  return certificate
+  note over Admin, E_Manager #FFaa00
+    -Org_1 takes "ownership" of this enclave,
+    -the certificate should be made available to the chaincode endorsement plugin,
+     so to include it in the responses (for eventual validation) 
+  end note
+end group
 
+group enclave registration
   activate Admin
+  Admin -> E_Manager ++: register
+  E_Manager -> ERCC1 ++: register Credentials
+  E_Manager -> ERCC2 : register Credentials
 
-  Admin -> ERCC1 ++: register (CC_NAME, CC_VER, MSP_ID_ORG_1, SIG_Org_1, ECC_S_PK, ECC_E_PK, ATTESTATION)
-  group verify and store
-
-    ERCC1 -> ERCC1 : check attestation signature & status
-    note left
-      - IAS verification key defined in ERCC definition
-      - status denotes the attestation result
-    end note
-
-    ERCC1 -> ERCC1 : Retrieve ECC definition for <CC_NAME, CC_VER>
+  ref over ERCC2
+    (same as ERCC1)
+  end ref
+  activate ERCC2
 
-    ERCC1 -> ERCC1 : Verify SIG_Org_1
+  ERCC1 -> ERCC1 : check Credentials.Evidence
+  ERCC1 -> ERCC1 : check Credentials.Attested_Data.CC_Parameters
```

Comment:
Hmm, chaincode definition is not fabric internal secret, it is just not exposed to chaincode but is exposed to admins (which define it in the first place).   I still would prefer to understand right away all checks performed when looking at ercc code but if everybody else is happy with outsourcing the check rather than only the retrieval of the chaincode definition, then i will consider myself overruled :-)

Reply:
AFAIK, Fabric does not expose the chaincode definition data structure.
If this holds, then the chaincode definition **is** an internal secret and is **not** exposed to admins. In fact, the admins only specify some parameters, and it's the Fabric documentation that says that those are the chaincode definition.

Reply:
> I still would prefer to understand right away all checks performed when looking at ercc code

I really don't see how this would hinder your understanding, given that the one-line would be `if cc_def_exists(id, ver, seq) then ok, else abort`, where `cc_def_exists` is the call to TLCC.

> but if everybody else is happy with outsourcing the check rather than only the retrieval of the chaincode definition, then i will consider myself overruled 

If you mean "retrieving the protobuf", I would be very unhappy.
Otherwise... feel free to update it.

Reply:
> AFAIK, Fabric does not expose the chaincode definition data structure.

But we don't really care about the internal data structure, just the information which we also get from `peer lifecycle chaincode querycommitted` (which is probably a subset of the information in the channel definition?)

Reply:
> I really don't see how this would hinder your understanding, given that the one-line would be `if cc_def_exists(id, ver, seq) then ok, else abort`, where `cc_def_exists` is the call to TLCC.

If you rephrase it like that i would feel much better as it spells out the actual check.  It raises though another question: is checking for existence sufficient or should we check it is also currently _active_?   Issue probably primarily related to revocation? For dealing with confidentiality & revocation is a big rathole which also might require some re-encryption or more complicated mitigation/damage control but from an integrity perspective, probably purging pre-existing outdated enclaves from the registry and preventing any future registration of such outdated enclaves should be sufficient. So an "active-check" could be useful.

Reply:
regarding chaincode definition. Clients can get the contents of the chaincode definition in human readable form by invoking lifecycle commands.

See example here: https://hyperledger-fabric.readthedocs.io/en/release-2.0/commands/peerlifecycle.html#peer-lifecycle-chaincode-querycommitted-example

The corresponding proto message you can find here: https://github.com/hyperledger/fabric-protos/blob/9944f8f73cfe0585a477f4b7d56a090374d0c7b6/peer/lifecycle/lifecycle.proto#L173

As a system chaincode we have the chance to retrieve a chaincode definition from the peer by following the implementation of the lifecycle command, see here: https://github.com/hyperledger/fabric/blob/master/core/chaincode/lifecycle/scc.go#L421

Some more references
- https://github.com/hyperledger/fabric/blob/9277b0eb3e6c4dadc37c69602e8e8f9c14af36ab/core/chaincode/lifecycle/scc.go#L95
- https://github.com/hyperledger/fabric/blob/9277b0eb3e6c4dadc37c69602e8e8f9c14af36ab/core/chaincode/lifecycle/scc.go#L421

Reply:
> regarding chaincode definition. Clients/Admins can get the contents of the approved chaincode definition in human readable form by invoking lifecycle commands.

That's what i wrote just two comments above in the same thread ;-)
 
> The corresponding proto message you can find here: https://github.com/hyperledger/fabric-protos/blob/9944f8f73cfe0585a477f4b7d56a090374d0c7b6/peer/lifecycle/lifecycle.proto#L173

Note this though almost certainly _not_ the chaincode definition on the ledger: there is no code notarized on the ledger with Fabric v2.  That said, it does contain the information which we care about and that part should correspond to what is on the ledger.

Reply:
> That's what i wrote just two comments above in the same thread ;-)

so I just confirmed what you wrote! :)